dcpquery/etl/__init__.py

```diff
@@ -36,21 +32,83 @@ def transform_bundle(bundle_uuid, bundle_version, bundle_path, bundle_manifest_p
-    bf_links = []
-    bundle_row = Bundle(uuid=bundle["uuid"], version=bundle["version"], manifest=bundle["manifest"])
-    for f in bundle["files"]:
-        filename = f.pop("name")
-        if filename == "links.json":
-            links = f['body']['links']
-        file_row = File(**f)
-        bf_links.append(BundleFileLink(bundle=bundle_row, file=file_row, name=filename))
-    config.db_session.add_all(bf_links)
-    config.db_session.commit()
-    load_links(links)
+def combine_file_data(bundle):
+    combined_file_data = {}
+    for f in bundle["manifest"]['files']:
+        uuid = f['uuid']
+        combined_file_data[uuid] = f
+
+    for f in bundle['files']:
+        uuid = f['uuid']
+        combined_file_data[uuid] = {**f, **combined_file_data[uuid]}
+    for uuid in combined_file_data:
+        file = combined_file_data[uuid]
+        filename = file['name']
+        regex_search_result = re.search('^(?:(?:.){1,}?)((?:[.][a-z]{1,10}){1,2})$', filename)
+        if regex_search_result:
+            file_extension = regex_search_result.group(1)
+        if 'body' not in file:
+            file['body'] = {'schema_type': None}
+        file['file_extension'] = file_extension or None
+    bundle['combine_file_data'] = combined_file_data
+
+
+class BundleLoad:
+    def __init__(self):
+        schema_types = config.db_session.query(SchemaType).with_entities(SchemaType.name).all()
+        self.schema_types = [schema[0] for schema in schema_types]
+
+    def schema_type_check(self, schema_type):
+        if schema_type and schema_type not in self.schema_types:
+            schema = SchemaType(name=schema_type)
+            config.db_session.add(schema)
+            config.db_session.commit()
+            self.schema_types.append(schema_type)
+
+    def load_bundle(self, bundle, extractor, transformer):
+        bf_links = []
+        bundle_row = Bundle(uuid=bundle["uuid"], version=bundle["version"], manifest=bundle["manifest"])
+
+        for uuid in bundle['combine_file_data']:
+            file_data = bundle['combine_file_data'][uuid]
+            filename = file_data.pop("name")
+            file_extension = file_data.pop("file_extension")
+
+            if filename == "links.json":
+                links = file_data['body']['links']
+                load_links(links)
+
+            self.schema_type_check(file_data['body']['schema_type'])
+            file_row = File(
+                uuid=file_data['uuid'],
+                version=file_data['version'],
+                content_type=file_data['content-type'],
+                size=file_data['size'],
+                extension=str(file_extension),
+                body=file_data['body'],
+                schema_type_name=file_data['body']['schema_type']
+
+            )
+
+            bf_links.append(BundleFileLink(bundle=bundle_row, file=file_row, name=filename))
+        config.db_session.add_all(bf_links)
+        config.db_session.commit()
+        self.create_view_tables()
```

Comment:
Would it be possible to query the database for all unique schema types and create them in the finalizer instead, instead of the loader? This might avoid a serious performance overhead in the loader.

Reply:
👍